packages/file-search/src/node/file-search-service-impl.ts

```diff
@@ -68,35 +68,45 @@ export class FileSearchServiceImpl implements FileSearchService {
             }
         }
 
-        const exactMatches = new Set<string>();
-        const fuzzyMatches = new Set<string>();
+        const matches = new Set<string>();
+
         const stringPattern = searchPattern.toLocaleLowerCase();
         await Promise.all(Object.keys(roots).map(async root => {
             try {
                 const rootUri = new URI(root);
                 const rootOptions = roots[root];
-                await this.doFind(rootUri, rootOptions, candidate => {
-                    const fileUri = rootUri.resolve(candidate).toString();
-                    if (exactMatches.has(fileUri) || fuzzyMatches.has(fileUri)) {
-                        return;
-                    }
-                    if (!searchPattern || searchPattern === '*' || candidate.toLocaleLowerCase().indexOf(stringPattern) !== -1) {
-                        exactMatches.add(fileUri);
-                    } else if (opts.fuzzyMatch && fuzzy.test(searchPattern, candidate)) {
-                        fuzzyMatches.add(fileUri);
-                    }
-                    if (exactMatches.size + fuzzyMatches.size === opts.limit) {
-                        cancellationSource.cancel();
-                    }
-                }, token);
+                // Perform searches for `exact` and `fuzzy` matches in parallel.
+                await Promise.all([
```

Comment:
Why do we need to run the same command twice in parallel?

Reply:
@akosyakov 

The initial idea was that I wanted to run the `exact` matches first to try to fill up the limit.
Exact matches will give us the best possible results from a user standpoint. If there would be remaining space (the limit was not reached), then the remaining space would be dedicated to `fuzzy` matches. 

The initial problem was that we checked at the same time if a result is an `exact` match and if not is it a `fuzzy` match leading to the limit being hit with much more fuzzy matches (and never seeing much better `exact` matches later on).

After @kittaakos https://github.com/theia-ide/theia/pull/5638#issuecomment-509616237, I tried to perform these searches in parallel to squeeze out any performance I could.

Please let me know if something can be optimized or if anything needs addressing better.

Reply:
> The initial idea was

It's all clear.

But since the two `doFind` runs in asynchronously, how can you ensure that the exact matches will fill the array first? Then comes the rest with the fuzzy match? Perhaps my ripgrep knowledge is limited on this :/

Reply:
> But since the two doFind runs in asynchronously, how can you ensure that the exact matches will fill the array first?

I can update the code to go back to using two sets (`exactMatches` and `fuzzyMatches`) to ensure each search goes to their respective set.

> Perhaps my ripgrep knowledge is limited on this...

Mine is as well :(